### PR TITLE
[FEATURE] Implement Spark Decorators and Helpers; Demonstrate on MulticolumnSumEqual Metric

### PR DESCRIPTION
**Pull request by alexsherstinsky**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: GREAT-56/GREAT-13
- JIRA: GREAT-56/GREAT-29
- JIRA: GREAT-56/GREAT-17
- JIRA: GREAT-56/GREAT-107
- JIRA: GREAT-56/GREAT-114
- JIRA: GREAT-56/GREAT-123
- JIRA: GREAT-56/GREAT-127

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [ ] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [ ] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [ ] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/expectations/metrics/map_metric_provider.py [hunk 1]
```diff
@@ -925,15 +1051,76 @@ def inner_func(
...
+
+            return inner_func
+
+        return wrapper
+
+    elif issubclass(engine, SparkDFExecutionEngine):
+        if partial_fn_type is None:
+            partial_fn_type = MetricPartialFunctionTypes.MAP_FN
+        partial_fn_type = MetricPartialFunctionTypes(partial_fn_type)
+        if partial_fn_type != MetricPartialFunctionTypes.MAP_FN:
```
Comment (fjork3) [inline great_expectations/expectations/metrics/map_metric_provider.py:1076]: style nit: line 1282 uses `not in` for checking the function type; this seems like it should be matched to that.

File: great_expectations/expectations/metrics/map_metric_provider.py [hunk 2]
```diff
@@ -806,6 +867,72 @@ def inner_func(
...
+    elif issubclass(engine, SparkDFExecutionEngine):
+        if partial_fn_type is None:
+            partial_fn_type = MetricPartialFunctionTypes.MAP_CONDITION_FN
+        partial_fn_type = MetricPartialFunctionTypes(partial_fn_type)
+        if partial_fn_type not in [
+            MetricPartialFunctionTypes.MAP_CONDITION_FN,
+            MetricPartialFunctionTypes.WINDOW_CONDITION_FN,
+        ]:
+            raise ValueError(
+                "SparkDFExecutionEngine only supports map_condition_fn for column_pair_condition_partial partial_fn_type"
```
Comment (fjork3) [inline great_expectations/expectations/metrics/map_metric_provider.py]: From the above block, it looks like `window_condition_fn` is also supported.

File: great_expectations/expectations/metrics/map_metric_provider.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/metrics/map_metric_provider.py:1076]: @fjork3 Good point -- we want to keep it as `!=` for now just to indicate that there can only be one.

File: great_expectations/expectations/metrics/map_metric_provider.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/metrics/map_metric_provider.py]: @fjork3 Great catch -- updated!  Thank you!

**Review by fjork3 — APPROVED** [review]
LGTM, thanks Alex!

**Pull request merged**